File: services/rag/app/services/search_service.py
"""Hybrid search service for the RAG pipeline.

BM25 full-text (pg_search) + pgvector similarity with RRF fusion.
Per-tenant scoping by `org_slug` (always applied), optional further
restriction by `file_ids` and a hierarchical `folder_path` prefix.
Optional semantic caching and cross-encoder re-ranking.
"""

from __future__ import annotations

import asyncio
import json
import time
from typing import Any, ClassVar

import asyncpg
from loguru import logger
from tale_knowledge.embedding import EmbeddingService, EmbeddingUsage
from tale_knowledge.retrieval import merge_rrf
from tale_knowledge.retrieval.reranker import Reranker
from tale_shared.db import acquire_with_retry

from ..config import settings
from .semantic_cache import SemanticCache

SCHEMA = "private_knowledge"


async def _timed(name: str, coro: Any) -> Any:
    t = time.time()
    try:
        return await coro
    finally:
        logger.debug("PERF {}: {:.1f}ms", name, (time.time() - t) * 1000)


class RagSearchService:
    _background_tasks: ClassVar[set[asyncio.Task[None]]] = set()

    def __init__(self, pool: asyncpg.Pool, embedding_service: EmbeddingService):
        self._pool = pool
        self._embedding = embedding_service
        self._semantic_cache: SemanticCache | None = SemanticCache(pool) if settings.semantic_cache_enabled else None
        self._reranker: Reranker | None = (
            Reranker(
                model_name=settings.reranking_model,
                provider=settings.reranking_provider,
            )
            if settings.reranking_enabled
            else None
        )

    async def search(
        self,
        org_slug: str,
        query: str,
        *,
        file_ids: list[str] | None = None,
        folder_path: str | None = None,
        top_k: int = 10,
        similarity_threshold: float = 0.0,
    ) -> tuple[list[dict[str, Any]], EmbeddingUsage]:
        """Hybrid BM25 + vector search scoped to `org_slug`.

        Args:
            org_slug: Tenant slug — ALWAYS used to filter `chunks.org_slug`
                so no cross-tenant rows can match.
            query: Search query text.
            file_ids: Optional file IDs to further restrict search to. The
                org filter is independent — file_ids only narrows within
                the org.
            folder_path: Optional hierarchical folder prefix filter
                (normalized, no surrounding slashes). Pure narrowing within
                the already-authorized file_ids scope — never an
                authorization boundary.
            top_k: Maximum number of results to return.
            similarity_threshold: Minimum cosine similarity for vector results.
                Results below this threshold are discarded before RRF merge.

        Returns:
            ``(results, embedding_usage)`` — usage returned alongside the
            results so concurrent callers can't trample each other's
            attribution via a shared singleton.
        """
        query_embedding: list[float] | None = None
        usage = EmbeddingUsage(model=self._embedding._model)
        try:
            t0 = time.time()
            query_result, fts_results = await asyncio.gather(
                _timed("embed", self._embedding.embed_query_with_usage(query)),
                _timed(
                    "fts",
                    self._fts_search(query, org_slug, file_ids, top_k * 3, folder_path=folder_path),
                ),
            )
            query_embedding = query_result.embedding
            usage = query_result.usage
            logger.debug("PERF embed+FTS total: {:.1f}ms", (time.time() - t0) * 1000)

            # Semantic cache: check for a cached result before vector search.
            # Cache is org-scoped — see SemanticCache.lookup. Bypassed (both
            # lookup and store) for folder-filtered queries: the cache key is
            # (org_slug, embedding) only, so a folder-scoped query would hit
            # unfiltered cached results and vice versa.
            if self._semantic_cache and query_embedding and not folder_path:
                cache_t0 = time.time()
                cached = await self._semantic_cache.lookup(
                    org_slug,
                    query_embedding,
                    threshold=settings.semantic_cache_similarity_threshold,
                )
                cache_ms = (time.time() - cache_t0) * 1000
                if cached:
                    logger.debug("Semantic cache hit for query (lookup {:.1f}ms): {}", cache_ms, query[:80])
                    try:
                        cached_results = json.loads(cached.response_text)
                        for r in cached_results:
                            r["cached"] = True
                        return cached_results, usage
                    except (json.JSONDecodeError, TypeError):
                        logger.warning("Invalid cached response format, performing fresh search")

            vec_t0 = time.time()
            vector_results = await self._vector_search(
                query_embedding, org_slug, file_ids, top_k * 3, folder_path=folder_path
            )
            vec_ms = (time.time() - vec_t0) * 1000
            logger.debug("PERF vector search: {:.1f}ms", vec_ms)

            # Pre-filter vector results by cosine similarity to reject clearly irrelevant content.
            # If ALL vector results are below threshold, the query is semantically irrelevant
            # to the indexed documents — discard FTS results too (they are keyword noise).
            if similarity_threshold > 0:
                pre_count = len(vector_results)
                top_score = max((r["score"] for r in vector_results), default=0.0)
                vector_results = [r for r in vector_results if r["score"] >= similarity_threshold]
                if pre_count != len(vector_results):
                    logger.debug(
                        "Vector pre-filter: {}/{} results passed threshold {} (top score {:.3f})",
                        len(vector_results),
                        pre_count,
                        similarity_threshold,
                        top_score,
                    )
                if pre_count > 0 and not vector_results:
                    return [], usage

            if not fts_results and not vector_results:
                return [], usage

            merged = merge_rrf([fts_results, vector_results], top_k)

            if settings.recency_boost_enabled:
                _apply_recency_boost(
                    merged,
                    decay_base=settings.recency_decay_base,
                    max_age_days=settings.recency_max_age_days,
                )

            # Re-rank merged results with cross-encoder if enabled.
            # The reranker's input "content" and the returned payload "content"
            # both prefer `core_content` (the chunk's non-overlap forward-owning
            # span), falling back to `chunk_content` for un-reindexed rows.
            # This prevents adjacent search hits from duplicating overlap bytes
            # to the LLM once Part B Phase 3 reindex completes.
            if self._reranker and merged:
                rerank_t0 = time.time()
                rerank_input = [
                    {"content": (item.get("core_content") or item.get("chunk_content") or ""), **item}
                    for item in merged
                ]
                merged = await self._reranker.rerank(
                    query,
                    rerank_input,
                    top_k=settings.reranking_top_k,
                )
                rerank_ms = (time.time() - rerank_t0) * 1000
                logger.debug("PERF reranking: {:.1f}ms", rerank_ms)

            results = [
                {
                    "content": item.get("core_content") or item.get("chunk_content") or "",
                    "score": item.get("reranking_score", item["rrf_score"]),
                    "file_id": str(item["file_id"]) if item.get("file_id") else None,
                    "filename": item.get("filename"),
                    "source_created_at": item.get("source_created_at"),
                    "source_modified_at": item.get("source_modified_at"),
                }
                for item in merged
            ]

            # Semantic cache: store results for future lookups (org-scoped).
            # Folder-filtered results are never stored — see the lookup bypass.
            if self._semantic_cache and query_embedding and results and not folder_path:
                result_file_ids = [r["file_id"] for r in results if r.get("file_id")]
                await self._semantic_cache.store(
                    org_slug,
                    query,
                    query_embedding,
                    json.dumps(results, default=str),
                    ttl_hours=settings.semantic_cache_ttl_hours,
                    file_ids=result_file_ids,
                )

            return results, usage

        except asyncpg.UndefinedTableError:
            logger.info("Tables not yet created, returning empty results")
            return [], usage
        except asyncpg.UndefinedColumnError:
            logger.info("Schema not ready, returning empty results")
            return [], usage
        except (asyncpg.InternalServerError, asyncpg.DataCorruptedError) as e:
            is_bm25 = "bm25" in str(e).lower()
            is_corruption = isinstance(e, asyncpg.DataCorruptedError)

            if is_bm25 or is_corruption:
                logger.warning("BM25 index issue (corruption={}): {}, falling back to vector-only", is_corruption, e)

                if is_corruption:
                    task = asyncio.create_task(self._rebuild_bm25_index())
                    self._background_tasks.add(task)
                    task.add_done_callback(self._background_tasks.discard)

                if query_embedding is None:
                    query_embedding = await self._embedding.embed_query(query)
                vector_results = await self._vector_search(
                    query_embedding, org_slug, file_ids, top_k, folder_path=folder_path
                )
                results = [
                    {
                        "content": item.get("core_content") or item.get("chunk_content") or "",
                        "score": item["score"],
                        "file_id": str(item["file_id"]) if item.get("file_id") else None,
                        "filename": item.get("filename"),
                        "source_created_at": item.get("source_created_at"),
                        "source_modified_at": item.get("source_modified_at"),
                    }
                    for item in vector_results
                ]
                return results, usage
            raise

    def _build_scope_clause(
        self,
        org_slug: str,
        file_ids: list[str] | None,
        param_offset: int,
        folder_path: str | None = None,
    ) -> tuple[str, list[Any]]:
        """Build WHERE clause for per-tenant + optional per-document scoping.

        `org_slug` is ALWAYS added as `AND c.org_slug = $N`. Document-level
        filters (`file_ids`, `folder_path`) are collected into a condition
        list applied through a single documents subquery that is itself
        org-scoped (defense in depth — even if the outer chunks filter were
        ever bypassed by a code mistake, the inner documents lookup also has
        the org filter). #1517's generic metadata filters extend the same
        condition list.

        `folder_path` is a boundary-safe hierarchical prefix: it matches the
        folder itself and any descendant ('data-room' matches
        'data-room/contracts' but NOT the sibling 'data-room-x'). The
        `left()` comparison avoids LIKE-pattern escaping for user-supplied
        paths; it is not index-sargable, which is acceptable because the
        subquery is already org- and file_ids-bounded.
        """
        org_param = param_offset + 1
        clause = f" AND c.org_slug = ${org_param}"
        params: list[Any] = [org_slug]

        doc_conditions: list[str] = []
        if file_ids:
            params.append(file_ids)
            doc_conditions.append(f"file_id = ANY(${param_offset + len(params)})")
        if folder_path:
            params.append(folder_path)
            folder_param = f"${param_offset + len(params)}"
            doc_conditions.append(
                f"(folder_path = {folder_param} "
                f"OR left(folder_path, char_length({folder_param}) + 1) = {folder_param} || '/')"
            )

        if doc_conditions:
            clause += (
                f" AND c.document_id IN ("
                f"SELECT id FROM {SCHEMA}.documents "
                f"WHERE org_slug = ${org_param} AND {' AND '.join(doc_conditions)})"
            )

        return clause, params

    async def _rebuild_bm25_index(self) -> None:
        """Rebuild the BM25 index after corruption. Runs as a background task."""
        try:
            logger.warning("Rebuilding BM25 index due to corruption")
            async with acquire_with_retry(self._pool) as conn:
                await conn.execute(f"REINDEX INDEX {SCHEMA}.idx_pk_chunks_bm25")
            logger.info("BM25 index rebuilt successfully")
        except Exception as e:
            logger.error("BM25 index rebuild failed: {}", e)

    async def _fts_search(
        self,
        query: str,
        org_slug: str,
        file_ids: list[str] | None,
        limit: int,
        folder_path: str | None = None,
    ) -> list[dict[str, Any]]:
        tenant_clause, tenant_params = self._build_scope_clause(org_slug, file_ids, 1, folder_path=folder_path)

        sql = f"""
            SELECT c.id, c.chunk_content, c.core_content, c.chunk_index, c.document_id,
                   d.file_id, d.filename,
                   d.source_created_at, d.source_modified_at, d.created_at,
                   paradedb.score(c.id) AS score
            FROM {SCHEMA}.chunks c
            LEFT JOIN {SCHEMA}.documents d ON c.document_id = d.id
            WHERE c.id @@@ paradedb.match('chunk_content', $1)
            {tenant_clause}
            ORDER BY score DESC
            LIMIT ${2 + len(tenant_params)}
        """
        params = [query, *tenant_params, limit]

        try:
            async with acquire_with_retry(self._pool) as conn:
                rows = await conn.fetch(sql, *params)
                return [dict(r) for r in rows]
        except asyncpg.DataCorruptedError as e:
            logger.warning("BM25 index corrupted: {}", e)
            return []
        except asyncpg.InternalServerError as e:
            logger.warning("FTS search failed: {}", e)
            return []

    async def _vector_search(
        self,
        embedding: list[float],
        org_slug: str,
        file_ids: list[str] | None,
        limit: int,
        folder_path: str | None = None,
    ) -> list[dict[str, Any]]:
        vec_str = json.dumps(embedding)
        tenant_clause, tenant_params = self._build_scope_clause(org_slug, file_ids, 1, folder_path=folder_path)

        sql = f"""
            SELECT c.id, c.chunk_content, c.core_content, c.chunk_index, c.document_id,
                   d.file_id, d.filename,
                   d.source_created_at, d.source_modified_at, d.created_at,
                   1 - (c.embedding <=> $1::vector) AS score
            FROM {SCHEMA}.chunks c
            LEFT JOIN {SCHEMA}.documents d ON c.document_id = d.id
            WHERE c.embedding IS NOT NULL
            {tenant_clause}
            ORDER BY c.embedding <=> $1::vector
            LIMIT ${2 + len(tenant_params)}
        """
        params = [vec_str, *tenant_params, limit]

        async with acquire_with_retry(self._pool) as conn:
            rows = await conn.fetch(sql, *params)
            return [dict(r) for r in rows]


def _apply_recency_boost(
    results: list[dict[str, Any]],
    decay_base: float,
    max_age_days: int,
) -> None:
    """Scale RRF scores by document age so newer documents rank higher.

    Modifies *results* in place: adjusts ``rrf_score``, re-normalises so the
    top result equals 1.0, and re-sorts descending.
    """
    from datetime import UTC, datetime

    now = datetime.now(UTC)
    for item in results:
        doc_ts = item.get("source_modified_at") or item.get("created_at")
        if doc_ts is None:
            item["rrf_score"] *= decay_base
            continue
        age_days = (now - doc_ts).total_seconds() / 86400
        recency_factor = max(0.0, 1.0 - age_days / max_age_days)
        boost = decay_base + (1.0 - decay_base) * recency_factor
        item["rrf_score"] *= boost

    results.sort(key=lambda x: x.get("rrf_score", 0), reverse=True)

File: services/platform/convex/rag/lib/search_service.ts
'use node';

/**
 * Hybrid search service for the RAG pipeline.
 *
 * BM25 full-text (pg_search) + pgvector similarity with RRF fusion. Per-tenant
 * scoping by `org_slug` is ALWAYS applied; optional further restriction by
 * `file_ids`, a hierarchical `folder_path` prefix, and flat document-metadata
 * filters. Optional semantic caching and cross-encoder re-ranking.
 */

import type { Sql } from 'postgres';
import { z } from 'zod';

import { withRetry } from '../../lib/knowledge/db/retry';
import type {
  EmbeddingService,
  EmbeddingUsage,
} from '../../lib/knowledge/embedding/service';
import { Reranker } from '../../lib/knowledge/retrieval/reranker';
import { mergeRrf } from '../../lib/knowledge/retrieval/rrf';

/** Values acceptable as positional parameters to postgres.js `sql.unsafe`. */
type SqlParam = string | number | boolean | null | Date | string[] | number[];

import {
  bm25Available,
  isDataCorrupted,
  isInternalError,
  isUndefinedColumn,
  isUndefinedFunction,
  isUndefinedSchema,
  isUndefinedTable,
  markBm25Unavailable,
  PRIVATE_KNOWLEDGE_SCHEMA as SCHEMA,
} from '../../lib/knowledge/db/knowledge_db';
import { logger } from '../../lib/knowledge/logger';
import { settings } from './config';
import { SemanticCache } from './semantic_cache';

export type MetadataFilterValue =
  | string
  | number
  | boolean
  | (string | number | boolean)[];

export interface SearchResultRow {
  content: string;
  score: number;
  file_id: string | null;
  filename: string | null;
  source_created_at: Date | null;
  source_modified_at: Date | null;
  cached?: boolean;
}

export interface SearchOptions {
  fileIds?: string[] | null;
  folderPath?: string | null;
  metadataFilters?: Record<string, MetadataFilterValue> | null;
  topK?: number;
  similarityThreshold?: number;
}

let sharedReranker: Reranker | null = null;

function getSharedReranker(): Reranker {
  if (sharedReranker === null) {
    sharedReranker = new Reranker({
      modelName: settings.reranking_model,
      provider: settings.reranking_provider === 'api' ? 'api' : 'local',
      apiBaseUrl: settings.reranking_api_base_url,
      apiKey: settings.reranking_api_key,
    });
  }
  return sharedReranker;
}

/** Render a scalar the way Postgres `jsonb ->> key` renders it. */
function jsonbScalarText(value: string | number | boolean): string {
  if (typeof value === 'string') {
    return value;
  }
  return JSON.stringify(value);
}

function emptyUsage(model: string): EmbeddingUsage {
  return { promptTokens: 0, totalTokens: 0, model };
}

// A type alias (not an interface) so it satisfies `mergeRrf`'s
// `Record<string, unknown>` constraint without an index signature poisoning
// property-access types.
type ChunkRow = {
  id: string;
  chunk_content: string | null;
  core_content: string | null;
  chunk_index: number;
  document_id: string;
  file_id: string | null;
  filename: string | null;
  source_created_at: Date | null;
  source_modified_at: Date | null;
  created_at: Date | null;
  score: number;
};

type MergedRow = ChunkRow & {
  rrf_score: number;
  reranking_score?: number;
};

// Rerank input drops the explicit nullable text fields and re-adds them as
// `string | undefined` so a MergedRow satisfies the reranker's
// `RerankableResult` contract. The reranker preserves all other fields.
type RerankRow = Omit<MergedRow, 'core_content' | 'chunk_content'> & {
  core_content?: string;
  chunk_content?: string;
  content: string;
};

const backgroundTasks = new Set<Promise<void>>();

export class RagSearchService {
  private readonly sql: Sql;
  private readonly embedding: EmbeddingService;
  private readonly semanticCache: SemanticCache | null;
  private readonly reranker: Reranker | null;

  constructor(sql: Sql, embedding: EmbeddingService) {
    this.sql = sql;
    this.embedding = embedding;
    this.semanticCache = settings.semantic_cache_enabled
      ? new SemanticCache(sql)
      : null;
    this.reranker = settings.reranking_enabled ? getSharedReranker() : null;
  }

  /** Build the WHERE clause for per-tenant + optional per-document scoping. */
  private buildScopeClause(
    orgSlug: string,
    fileIds: string[] | null | undefined,
    paramOffset: number,
    folderPath: string | null | undefined,
    metadataFilters: Record<string, MetadataFilterValue> | null | undefined,
  ): { clause: string; params: SqlParam[] } {
    const orgParam = paramOffset + 1;
    let clause = ` AND c.org_slug = $${orgParam}`;
    const params: SqlParam[] = [orgSlug];

    const docConditions: string[] = [];
    if (fileIds && fileIds.length > 0) {
      params.push(fileIds);
      docConditions.push(`file_id = ANY($${paramOffset + params.length})`);
    }
    if (folderPath) {
      params.push(folderPath);
      const folderParam = `$${paramOffset + params.length}`;
      docConditions.push(
        `(folder_path = ${folderParam} ` +
          `OR left(folder_path, char_length(${folderParam}) + 1) = ${folderParam} || '/')`,
      );
    }
    if (metadataFilters) {
      const equality: Record<string, string | number | boolean> = {};
      for (const [key, value] of Object.entries(metadataFilters)) {
        if (!Array.isArray(value)) {
          equality[key] = value;
        }
      }
      if (Object.keys(equality).length > 0) {
        params.push(JSON.stringify(equality));
        docConditions.push(
          `metadata @> $${paramOffset + params.length}::jsonb`,
        );
      }
      for (const [key, values] of Object.entries(metadataFilters)) {
        if (!Array.isArray(values)) {
          continue;
        }
        params.push(key);
        const keyParam = paramOffset + params.length;
        params.push(values.map((v) => jsonbScalarText(v)));
        docConditions.push(
          `metadata->>$${keyParam} = ANY($${paramOffset + params.length})`,
        );
      }
    }

    if (docConditions.length > 0) {
      clause +=
        ` AND c.document_id IN (` +
        `SELECT id FROM ${SCHEMA}.documents ` +
        `WHERE org_slug = $${orgParam} AND ${docConditions.join(' AND ')})`;
    }

    return { clause, params };
  }

  private async ftsSearch(
    query: string,
    orgSlug: string,
    fileIds: string[] | null | undefined,
    limit: number,
    folderPath: string | null | undefined,
    metadataFilters: Record<string, MetadataFilterValue> | null | undefined,
  ): Promise<ChunkRow[]> {
    // A database without pg_search (any managed Postgres) cannot run the
    // `paradedb.*` SQL below — skip the BM25 leg entirely so the search
    // degrades to vector-only instead of throwing (#2755).
    if (!(await bm25Available(this.sql))) {
      return [];
    }
    const { clause, params } = this.buildScopeClause(
      orgSlug,
      fileIds,
      1,
      folderPath,
      metadataFilters,
    );
    const sqlText = `
      SELECT c.id, c.chunk_content, c.core_content, c.chunk_index, c.document_id,
             d.file_id, d.filename,
             d.source_created_at, d.source_modified_at, d.created_at,
             paradedb.score(c.id) AS score
      FROM ${SCHEMA}.chunks c
      LEFT JOIN ${SCHEMA}.documents d ON c.document_id = d.id
      WHERE c.id @@@ paradedb.match('chunk_content', $1)
      ${clause}
      ORDER BY score DESC
      LIMIT $${2 + params.length}
    `;
    const allParams = [query, ...params, limit];
    try {
      return await withRetry(() =>
        this.sql.unsafe<ChunkRow[]>(sqlText, allParams),
      );
    } catch (err) {
      if (isDataCorrupted(err)) {
        logger.warn(`BM25 index corrupted: ${errMsg(err)}`);
        return [];
      }
      if (isInternalError(err)) {
        logger.warn(`FTS search failed: ${errMsg(err)}`);
        return [];
      }
      if (isUndefinedSchema(err) || isUndefinedFunction(err)) {
        // pg_search vanished (or the capability probe raced a DROP): remember
        // the answer so later searches skip the leg without erroring again.
        markBm25Unavailable(this.sql);
        logger.warn(
          `BM25 unavailable on this database, falling back to vector-only: ${errMsg(err)}`,
        );
        return [];
      }
      throw err;
    }
  }

  private async vectorSearch(
    embedding: number[],
    orgSlug: string,
    fileIds: string[] | null | undefined,
    limit: number,
    folderPath: string | null | undefined,
    metadataFilters: Record<string, MetadataFilterValue> | null | undefined,
  ): Promise<ChunkRow[]> {
    const vecStr = JSON.stringify(embedding);
    const { clause, params } = this.buildScopeClause(
      orgSlug,
      fileIds,
      1,
      folderPath,
      metadataFilters,
    );
    const sqlText = `
      SELECT c.id, c.chunk_content, c.core_content, c.chunk_index, c.document_id,
             d.file_id, d.filename,
             d.source_created_at, d.source_modified_at, d.created_at,
             1 - (c.embedding <=> $1::vector) AS score
      FROM ${SCHEMA}.chunks c
      LEFT JOIN ${SCHEMA}.documents d ON c.document_id = d.id
      WHERE c.embedding IS NOT NULL
      ${clause}
      ORDER BY c.embedding <=> $1::vector
      LIMIT $${2 + params.length}
    `;
    const allParams = [vecStr, ...params, limit];
    return withRetry(() => this.sql.unsafe<ChunkRow[]>(sqlText, allParams));
  }

  private async rebuildBm25Index(): Promise<void> {
    try {
      logger.warn('Rebuilding BM25 index due to corruption');
      await withRetry(() =>
        this.sql.unsafe(`REINDEX INDEX ${SCHEMA}.idx_pk_chunks_bm25`),
      );
      logger.info('BM25 index rebuilt successfully');
    } catch (err) {
      logger.error(`BM25 index rebuild failed: ${errMsg(err)}`);
    }
  }

  /** Hybrid BM25 + vector search scoped to `org_slug`. */
  async search(
    orgSlug: string,
    query: string,
    options: SearchOptions = {},
  ): Promise<[SearchResultRow[], EmbeddingUsage]> {
    const fileIds = options.fileIds ?? null;
    const folderPath = options.folderPath ?? null;
    const metadataFilters = options.metadataFilters ?? null;
    const topK = options.topK ?? 10;
    const similarityThreshold = options.similarityThreshold ?? 0.0;

    let queryEmbedding: number[] | null = null;
    let usage: EmbeddingUsage = emptyUsage('');

    try {
      const [queryResult, ftsResults] = await Promise.all([
        this.embedding.embedQueryWithUsage(query),
        this.ftsSearch(
          query,
          orgSlug,
          fileIds,
          topK * 3,
          folderPath,
          metadataFilters,
        ),
      ]);
      queryEmbedding = queryResult.embedding;
      usage = queryResult.usage;

      // Semantic cache lookup (org-scoped; bypassed for filtered queries).
      if (
        this.semanticCache &&
        queryEmbedding &&
        !folderPath &&
        !metadataFilters
      ) {
        const cached = await this.semanticCache.lookup(
          orgSlug,
          queryEmbedding,
          settings.semantic_cache_similarity_threshold,
        );
        if (cached) {
          logger.debug(`Semantic cache hit for query: ${query.slice(0, 80)}`);
          const parsedCache = parseCachedResults(cached.responseText);
          if (parsedCache !== null) {
            for (const r of parsedCache) {
              r.cached = true;
            }
            return [parsedCache, usage];
          }
          logger.warn(
            'Invalid cached response format, performing fresh search',
          );
        }
      }

      let vectorResults = await this.vectorSearch(
        queryEmbedding,
        orgSlug,
        fileIds,
        topK * 3,
        folderPath,
        metadataFilters,
      );

      // Pre-filter vector results by cosine similarity.
      if (similarityThreshold > 0) {
        const preCount = vectorResults.length;
        vectorResults = vectorResults.filter(
          (r) => r.score >= similarityThreshold,
        );
        if (preCount > 0 && vectorResults.length === 0) {
          return [[], usage];
        }
      }

      if (ftsResults.length === 0 && vectorResults.length === 0) {
        return [[], usage];
      }

      const rrfPoolSize = this.reranker
        ? Math.max(topK, settings.reranking_candidates)
        : topK;
      const merged: MergedRow[] = mergeRrf<ChunkRow>(
        [ftsResults, vectorResults],
        rrfPoolSize,
      );

      if (settings.recency_boost_enabled) {
        applyRecencyBoost(
          merged,
          settings.recency_decay_base,
          settings.recency_max_age_days,
        );
      }

      // Ranked rows carry a `content` field + an optional `reranking_score`;
      // the reranker, when enabled, reorders + scores them. Built with a
      // for-loop + `Object.assign` (no `map` + spread, which the no-map-spread
      // rule flags) — the assign writes into a fresh target so each
      // `MergedRow` is left untouched.
      let rankedRows: RerankRow[] = [];
      for (const item of merged) {
        const row: RerankRow = Object.assign({}, item, {
          core_content: item.core_content ?? undefined,
          chunk_content: item.chunk_content ?? undefined,
          content: item.core_content || item.chunk_content || '',
        });
        rankedRows.push(row);
      }

      if (this.reranker && rankedRows.length > 0) {
        try {
          rankedRows = await this.reranker.rerank<RerankRow>(
            query,
            rankedRows,
            Math.min(topK, settings.reranking_top_k),
          );
        } catch (rerankErr) {
          logger.warn(
            `Re-ranking failed, falling back to RRF order: ${errMsg(rerankErr)}`,
          );
          rankedRows = rankedRows.slice(0, topK);
        }
      } else {
        rankedRows = rankedRows.slice(0, topK);
      }

      const results: SearchResultRow[] = rankedRows.map((item) => ({
        content: item.core_content || item.chunk_content || '',
        score: item.reranking_score ?? item.rrf_score,
        file_id: item.file_id ? item.file_id : null,
        filename: item.filename ?? null,
        source_created_at: item.source_created_at ?? null,
        source_modified_at: item.source_modified_at ?? null,
      }));

      // Semantic cache store (org-scoped; never for filtered results).
      if (
        this.semanticCache &&
        queryEmbedding &&
        results.length > 0 &&
        !folderPath &&
        !metadataFilters
      ) {
        const resultFileIds = results
          .map((r) => r.file_id)
          .filter((id): id is string => Boolean(id));
        await this.semanticCache.store(
          orgSlug,
          query,
          queryEmbedding,
          JSON.stringify(results),
          {
            ttlHours: settings.semantic_cache_ttl_hours,
            fileIds: resultFileIds,
          },
        );
      }

      return [results, usage];
    } catch (err) {
      if (isUndefinedTable(err)) {
        logger.info('Tables not yet created, returning empty results');
        return [[], usage];
      }
      if (isUndefinedColumn(err)) {
        logger.info('Schema not ready, returning empty results');
        return [[], usage];
      }

      const isBm25 =
        err instanceof Error && err.message.toLowerCase().includes('bm25');
      const isCorruption = isDataCorrupted(err);
      if ((isInternalError(err) || isCorruption) && (isBm25 || isCorruption)) {
        logger.warn(
          `BM25 index issue (corruption=${isCorruption}): ${errMsg(err)}, falling back to vector-only`,
        );
        if (isCorruption) {
          const task = this.rebuildBm25Index();
          backgroundTasks.add(task);
          void task.finally(() => backgroundTasks.delete(task));
        }
        if (queryEmbedding === null) {
          queryEmbedding = await this.embedding.embedQuery(query);
        }
        const vectorResults = await this.vectorSearch(
          queryEmbedding,
          orgSlug,
          fileIds,
          topK,
          folderPath,
          metadataFilters,
        );
        const results: SearchResultRow[] = vectorResults.map((item) => ({
          content: item.core_content || item.chunk_content || '',
          score: item.score,
          file_id: item.file_id ?? null,
          filename: item.filename,
          source_created_at: item.source_created_at,
          source_modified_at: item.source_modified_at,
        }));
        return [results, usage];
      }
      throw err;
    }
  }
}

function errMsg(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

const cachedResultSchema = z.object({
  content: z.string().default(''),
  score: z.number().default(0),
  file_id: z.string().nullable().default(null),
  filename: z.string().nullable().default(null),
  source_created_at: z.coerce.date().nullable().default(null),
  source_modified_at: z.coerce.date().nullable().default(null),
});

/** Parse a serialized cached result list, returning null on any malformation. */
function parseCachedResults(responseText: string): SearchResultRow[] | null {
  try {
    const parsed = z
      .array(cachedResultSchema)
      .safeParse(JSON.parse(responseText));
    return parsed.success ? parsed.data : null;
  } catch {
    return null;
  }
}

/** Scale RRF scores by document age so newer documents rank higher. */
export function applyRecencyBoost(
  results: MergedRow[],
  decayBase: number,
  maxAgeDays: number,
): void {
  const now = Date.now();
  for (const item of results) {
    const docTs = item.source_modified_at ?? item.created_at;
    const rrfScore = item.rrf_score ?? 0;
    if (docTs == null) {
      item.rrf_score = rrfScore * decayBase;
      continue;
    }
    const ageDays = (now - new Date(docTs).getTime()) / (86400 * 1000);
    const recencyFactor = Math.max(0.0, 1.0 - ageDays / maxAgeDays);
    const boost = decayBase + (1.0 - decayBase) * recencyFactor;
    item.rrf_score = rrfScore * boost;
  }
  results.sort((a, b) => (b.rrf_score ?? 0) - (a.rrf_score ?? 0));
}
